✨ Add deprecation date in backend/pgSQL

mrnicegyu11 opened the pull request:
<!-- Common title prefixes/annotations:
PREFIX:

  WIP: work in progress
  🐛    Fix a bug.
  ✨    Introduce new features.
  ♻️     Refactor code.
  🚑️    Critical hotfix.
  ⚗️     Perform experiments.
  ⬆️    Upgrade dependencies.
  📝    Add or update documentation.
  🗑️    Deprecate code that needs to be cleaned up.
  ⚰️     Remove dead code.
  🔥    Remove code or files.
  🔨    Add or update development scripts.

or from https://gitmoji.dev/ and https://github.com/carloscuesta/gitmoji/blob/master/src/data/gitmojis.json

SUFFIX:
 (⚠️ devops)  changes in devops configuration required before deploying

-->

## What do these changes do?

A column in the `service_meta_data` table is added to give a date from which on a service is considered deprecated. The field can be NULL or datetime. The default is NULL, a datetime can be manually added per service once it shall be deprecated. The `deprecated` column is passed to the frontend. For now, all further logic handling deprecation is done in the frontend. More information on service deprecation is found in the sprint planning slides.

Bonus: Fixes deprecated default credentials and typo in alembic migration readme.

## Related issue/s

<!-- Enumerate REVIEWERS other issues

- ITISFoundation/osparc-issues#428
- #26 : node_ports should have retry policies when upload/download fails  (FIXED)

-->


## How to test

<!-- Give REVIEWERS some hits or code snippets on how could this be tested -->


## Checklist

<!-- This is YOUR section

Add here YOUR checklist/notes to guide and monitor the progress of the case!

e.g.

- [ ] Openapi changes? ``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [ ] Database migration script? ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
- [ ] Unit tests for the changes exist
- [ ] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)
-->

pcrespov commented on packages/postgres-database/src/simcore_postgres_database/migration/versions/cf3bac482ce0_add_file_meta_data_deprecation_date.py:
```diff
@@ -0,0 +1,33 @@
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_column("services_meta_data", "deprecated")
```
did you forget to commit changes in packages/postgres-database/src/simcore_postgres_database/models/services.py ??

mrnicegyu11 replied:
sorry this was draft, thanks for the suggestion, changes are now added

sanderegg commented on packages/models-library/src/models_library/services_db.py:
```diff
@@ -34,6 +34,7 @@ class Config:
                 "thumbnail": "http://thumbnailit.org/image",
                 "created": "2021-01-18 12:46:57.7315",
                 "modified": "2021-01-19 12:45:00",
+                "deprecated": "2099-01-19 12:45:00",
```
well I guess it is a WIP PR right? do you really want comments as of now?
if deprecated is a datetime, then define it as such. using ```str``` is easier but it is also a great pain later since anything can be in there, in any format, etc...

mrnicegyu11 replied:
true, it was WIP sorry.

I cannot comment on pros/cons of str and datetime. I have switched to datetime now as per your suggestion.

sanderegg commented on packages/postgres-database/src/simcore_postgres_database/migration/README.md:
```diff
@@ -78,8 +80,8 @@ We create a revision script for the change by using the local db as follows:
 docker ps # find the published port for the db
-sc-pg discover -u simcore -p simcore --port=32787 # discover the db
-sp-pg info # what revision are we at?
+sc-pg discover -u scu -p adminadmin --port=5432 # discover the db
```
hem

mrnicegyu11 replied:
what does this mean?

sanderegg replied:
aren't you pushing some usernames-passwords here?

mrnicegyu11 replied:
Already pushed to upstream/master in .env-devel:
![image](https://user-images.githubusercontent.com/8209087/175051861-92ec0553-32be-4f8d-8307-309364d6f458.png)

Nothing is revealed that is not already there

sanderegg commented on packages/postgres-database/src/simcore_postgres_database/migration/versions/cf3bac482ce0_add_file_meta_data_deprecation_date.py:
```diff
@@ -0,0 +1,33 @@
+    op.add_column(
+        "services_meta_data",
+        sa.Column(
+            "deprecated", postgresql.TIMESTAMP(), autoincrement=False, nullable=True
```
ok I see this now for the first time, and I see @pcrespov used that already in the project snapshots. what is the advantage of timestamp over datetime?

mrnicegyu11 replied:
I cannot comment on pros/cons of str or datetime, sorry

sanderegg replied:
well then let's not mix them. can you check with @pcrespov why and if this should be used? since this is a copy paste of probably the only location where it was used... maybe let's check first if it has pros-cons

sanderegg replied:
from sqlalchemy [doc](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.DateTime) I can see this brings timezone support. which in itself might be a good idea... but still we do not use this anywhere. Maybe you could create an issue that we could convert to that format all our times in the database that currently use the standard python datetime

pcrespov replied:
All good!`` postgresql.TIMESTAMP()`` is auto-generated from ``sa.DateTime`` by alembic and represents a special type for postgres

sanderegg replied:
sorry but I disagree.
looking at all of the other migration scripts show sa.DateTime in them, not a postgresql.TIMESTAMP(). So let's have alook together why that is (if it's something new from alembic versions or an error). If possible I would like that we do not start mixing up types more than necessary

sanderegg reviewed (commented):
just add me as reviewer when this is complete, unless you want more comments as it goes.

GitHK commented on services/catalog/src/simcore_service_catalog/db/repositories/services.py:
```diff
@@ -203,13 +203,16 @@ async def create_service(
         async with self.db_engine.begin() as conn:
             # NOTE: this ensure proper rollback in case of issue
-            result = await conn.execute(
+            result = await conn.execute(  ## Here, maybe add default deprecation date?
```
If this is not a todo:
1. Why are you hesitating in adding it?
1. What is a default deprecation date, can you explain?

sanderegg commented on packages/models-library/src/models_library/services_db.py:
```diff
@@ -34,6 +35,7 @@ class Config:
                 "thumbnail": "http://thumbnailit.org/image",
                 "created": "2021-01-18 12:46:57.7315",
                 "modified": "2021-01-19 12:45:00",
+                "deprecated": datetime.fromisoformat("2099-01-19 12:45:00"),
```
not sure why you need to use datetime here... the example is usually directly a dict

mrnicegyu11 replied:
will change

sanderegg commented on packages/postgres-database/src/simcore_postgres_database/models/services.py:
```diff
@@ -69,6 +70,13 @@
     ),
+    sa.Column(
+        "deprecated",
+        sa.DateTime(),
```
here you use datetime... you probably need to re-generate the alembic migration code.

sanderegg reviewed (changes requested):
cool, but please ensure the migration code fits with the postgres declaration.
one uses sa.DateTime, the other sa.Timestamp

mrnicegyu11 commented:
@sanderegg I think I added this a comment but it vanished. The credentials I updated are taken from .env-devel, which is pushed to github anyway. So nothing is leaked.

@sanderegg I opened an issue for timezone support as requested but I am doubtful about the priority of this feature. Anyway: https://github.com/ITISFoundation/osparc-simcore/issues/3131

@sanderegg Pedro has been querried for the questions w.r.t. datetime and timestamp. But as said before, DateTime is just the sa type for what is called `timestamp` in pgsql

sanderegg reviewed (changes requested):
Please explain why all the other migration scripts use sa.DateTime instead of postgresql.TIMESTAMP and what are the implications.
Is usage exactly the same?

mrnicegyu11 commented:
@sanderegg harmonized the alembic datatye to sa.datetime

sanderegg commented on packages/postgres-database/src/simcore_postgres_database/migration/versions/cf3bac482ce0_add_file_meta_data_deprecation_date.py:
```diff
@@ -26,8 +25,6 @@ def upgrade():
-        sa.Column(
-            "deprecated", postgresql.TIMESTAMP(), autoincrement=False, nullable=True
-        ),
+        sa.Column("deprecated", sa.DateTime(), autoincrement=False, nullable=True),
```
cool, thanks!

sanderegg reviewed (approved):
:+1: thanks, and welcome in the commiters!
